keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,927 @@
+    // and referred to by the key's value in events and errors.
+    ReplicatedGangKey *ObjectFieldSelector
+
+    // Optional when GangMode = "ReplicatedGang".
+    // Forbidden otherwise.
+    Replicas int
+
+    // PodGroupSelector must be set when GangMode = "Single" or "Off".
+    // Must be one of:
+    //   - `metadata.labels['<KEY>']`
+    //   Note: Annotation not supported because it is not indexed easily, and it may not
+    //         have a form that can be printed without escaping.
+    // Each value of that key maps to a different gang, which will be anonymous in status,
+    // and referred to by the key's value in events and errors.
+    PodGroupSelector PodGroupSelector
+
+    // GangSchedulingPolicy defines the options applying to all pods in this gang.
+    // Forbidden if GangMode is set to "Off".
+    GangSchedulingPolicy GangSchedulingPolicy
+}
+
+
+type WorkloadStatus struct {
+  // Necessary status fields TBD.
+}
+```
+
+The individual `PodGroups` and `PodGroup` replicas are treated as independent gangs. As an example, if one of
+the groups can be scheduled and the other can't be - this is exactly what will happen. If the underlying
+user intention was to have either both of them or none of them running, they should form a single group and
+not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
+of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
+may consist of a number of such replicas.
+
+### Scheduler Changes
+
+The kube-scheduler will be watching for `Workload` objects (using informers) and will use them to map pods
+to and from their `Workload` objects.
+
+In the initial implementation, we expect users to create the `Workload` objects. In the next steps controllers
+will be updated to create an appropriate `Workload` objects themselves whenever they can appropriately infer
+the intention from the desired state.
+Note that given scheduling options are stored in the `Workload` object, pods linked to the `Workload`
+object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.
+
+The north star vision for gang scheduling implementation should satisfy the following requirements:
+
+1. Ensure that pods being part of a gang are not scheduled if all pods belonging to it can't be scheduled.
+2. Provide the "optimal enough" placement by considering all pods from a gang together.
+3. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by kube-scheduler.
+4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by different
+   schedulers.
+5. Avoid preemptions of already running pods if these will not allow to schedule higher priority gang
+   anyway.
+6. Support gang-level (or workload-level in general) level preemption.
```

Comment:
Is each group (in workload.PodGroups[*]) supposed to have pods with the same template? Or can a single group contain various pods. I guess the latter is true, but in that case, the gang-aware preemption might be tricker because each pod in the same pod group could have a different priority.

Reply:
Yes - we definitely want to allow for the later (pods with different templates within a single PodGroup).

That being said, I would say that it doesn't mean that these can really be arbitrary pods with different priorities. We should discuss how to achieve that in the best way (whether you need to define priority explicitly at PodGroup level, whether we infer it somehow from the pods (min? max?) or something else, but having a gang of pods that would have different priorities shouldn't be supported usecase.

Reply:
My personal opinion is that we need a new "preemptionPolicy" entity that will effectively mean that certain groups of pods should be treated as a whole for preemption purposes. That will also contain the way how priority of it should be computed. This makes is strictly additive feature and should address the concerns.

That being said I added a paragraph to describe that and added a beta graduation criteria around that.